### ISSUE TITLE
(@aws-cdk/aws-dynamodb): DynamoDb Replication: Cannot exceed quota for PoliciesPerRole: 10 

### ISSUE THREAD

**Issue by reinismu** [aws/aws-cdk#13671]
```
Cannot exceed quota for PoliciesPerRole: 10 (Service: AmazonIdentityManagement; Status Code: 409; Error Code: LimitExceeded; Request ID: 3bd4add0-36e4-44ad-8e98-947977c0a638; Proxy: null)
```

At some point from version 1.90.0 till 1.94.1 I couldn't deploy my global tables.

I'm a bit busy now and will try to add more info later.

### Environment

  - **CDK CLI Version  :**
  - **Framework Version:**
  - **Node.js Version:** 
  - **OS               :**
  - **Language (Version):** 

### Other

--- 

This is :bug: Bug Report

**Comment by jogold** [comment]
How many tables and replica regions do you have in your stack?

**Comment by reinismu** [comment]
@jogold I have 11 tables and 9 of them are replicated in another region as well

**Comment by jogold** [comment]
@skinny85 this is indeed #13300, there's a limit of 10 managed policies per role and we now create 2 managed policies per table... any suggestion?

**Comment by skinny85** [comment]
I just want to understand the exact numbers here...

From what I see, we create 2 Policies, yes, but we attach them to different Roles: https://github.com/aws/aws-cdk/blob/aa45edeafbf37c5638378cd92d0aa7cc954d426c/packages/%40aws-cdk/aws-dynamodb/lib/table.ts#L1473-L1474 . So, we should be able to create 10 replicas, and only the 11th one should cause this error...

@reinismu can you clarify your answer a little bit? Because I don't exactly follow. Can you show the `replicationRegions` that you use in your code, and how many of them?

Thanks,
Adam

**Comment by reinismu** [comment]
Sorry, missed the comment.

```
const episodesTable = new Table(this, `${PREFIX}-series`, {
            partitionKey: {
                name: 'seriesId',
                type: AttributeType.STRING,
            },
            tableName: `${PREFIX}-series`,
            billingMode: BillingMode.PAY_PER_REQUEST,
            replicationRegions: replicationRegions,
            removalPolicy: REMOVAL_POLICY,
            pointInTimeRecovery,
        });
```
I have in total 9 tables with replicationRegions
        
`replicationRegions` comes from environment variables. Now we use eu-central-1 and ap-southeast-1

**Comment by skinny85** [comment]
Sorry, to clarify: you have 9 separate `Table` instances, and each is replicated to 2 regions?

**Comment by reinismu** [comment]
Yes :)

**Comment by skinny85** [comment]
Interesting! I just tried this:

```ts
class ExampleStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        for (let i = 1; i <= 9; i++) {
            new dynamodb.Table(this, `Table${i}`, {
                partitionKey: {
                    name: 'Id',
                    type: dynamodb.AttributeType.STRING,
                },
                replicationRegions: ['eu-central-1', 'ap-southeast-1'],
                removalPolicy: cdk.RemovalPolicy.DESTROY,
            });
        }
    }
}

const app = new cdk.App();

new ExampleStack(app, 'ExampleStack');
```

and it deployed successfully!

Can you try it on your end? Maybe the error is somewhere else?

Thanks,
Adam

**Comment by reinismu** [comment]
Not sure what's wrong, but your example doesn't work for me.
```
Template format error: Unresolved resource dependencies [Table7Replicaeucentral1E0F0C4BF] in the Resources block of the template
```

Can you try increasing 9 to 15 or something. Just to check. 

I upgraded my main infra stack to `1.95.1` and the issue is still there. In total it tries to create 20 Managed policies

**Comment by skinny85** [comment]
Hmm, that's very concerning. That error shouldn't happen for a such a simple template.

Can you push to a GitHub repo the minimal reproduction that makes you see that error? I'm curious what is causing it.

**Comment by reinismu** [comment]
I figured it out, Tho more by luck than error message. My main region is `eu-central-1` so I can't have it in `replicationRegions`. 

I'm deploying the example project and will try to break it now

**Comment by reinismu** [comment]
Aha got it!

I dpeloyed first with `for (let i = 1; i <= 9; i++) {` and then did another deploy with `for (let i = 1; i <= 12; i++) {` and it broke.

Not sure if deploying it 2x does anything, but can try

**Comment by skinny85** [comment]
OK, I least was able to reproduce the `Unresolved resource dependencies` error that you got 😜.

**Comment by skinny85** [comment]
First part of the fix: https://github.com/aws/aws-cdk/pull/13889

**Comment by skinny85** [comment]
Yep, I was able to reproduce the original error with 11 Tables (not sure how you're getting it with just 9, but that's besides the point). The problem is the Roles for the Custom Resource framework handlers. We create a separate Policy for each replicated Table, and attach all of them to those Roles, here: https://github.com/aws/aws-cdk/blob/93dac4c699e5ad817203398a7caa26695a3d7f47/packages/%40aws-cdk/aws-dynamodb/lib/table.ts#L1473-L1474

This is a tough one... I think the only way to do this properly is to somehow count how many `Custom::DynamoDBReplica` resources we have in the Stack, and once that number exceeds 10, we need to create new handlers (and thus Roles) for the Lambdas the custom resource framework needs.

@reinismu for now, I would split the resources into multiple Stacks, so that no Stack has more than 10 Tables with replicas.

@jogold any ideas on how to handle this?

**Comment by skinny85** [comment]
I wonder whether the logic of counting to 10 can be nicely encapsulated in the `getOrCreate ()` method here: https://github.com/aws/aws-cdk/blob/93dac4c699e5ad817203398a7caa26695a3d7f47/packages/%40aws-cdk/aws-dynamodb/lib/replica-provider.ts#L28-L32 , so that not much of the client code has to change...

Also, I think I found another potential error with this feature: today, `ReplicaProvider` is a singleton in a Stack, except there is a `replicationTimeout` property that you can pass it when creating it. Pretty sure the `replicationTimeout` property will be completely ignored if set on a second table that has any global replicas in one Stack...

**Comment by jogold** [comment]
> Also, I think I found another potential error with this feature: today, `ReplicaProvider` is a singleton in a Stack, except there is a `replicationTimeout` property that you can pass it when creating it. Pretty sure the `replicationTimeout` property will be completely ignored if set on a second table that has any global replicas in one Stack...

You are correct. What do you suggest here? Should we throw if a different value is set on a second table? Include the value in the `uid`?

**Comment by jogold** [comment]
> I wonder whether the logic of counting to 10 can be nicely encapsulated in the `getOrCreate ()` method here:

Not a really a fan of this counting solution but I can't think of another one...

**Comment by reinismu** [comment]
Could ask CloudFormation team why there is this limit? Maybe it's not that hard to get rid of it.

**Comment by jogold** [comment]
> Could ask CloudFormation team why there is this limit? Maybe it's not that hard to get rid of it.

It's an IAM limit: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html

You can request to increase it to 20.

**Comment by reinismu** [comment]
Ohh I see, I guess no way around that then. 

Even if I request it to 20 there would still be someone who will encounter this issue :/

**Comment by skinny85** [comment]
> > Also, I think I found another potential error with this feature: today, `ReplicaProvider` is a singleton in a Stack, except there is a `replicationTimeout` property that you can pass it when creating it. Pretty sure the `replicationTimeout` property will be completely ignored if set on a second table that has any global replicas in one Stack...
> 
> You are correct. What do you suggest here? Should we throw if a different value is set on a second table? Include the value in the `uid`?

Probably including the value of `timeout` in the UID is the way to go here.

**Comment by skinny85** [comment]
> > I wonder whether the logic of counting to 10 can be nicely encapsulated in the `getOrCreate ()` method here:
> 
> Not a really a fan of this counting solution but I can't think of another one...
> 
> Implementing it in `getOrCreate()` is just a matter of adding `Math.floor(counter/10)` in the `uid`?

@joggold is there any way we can do a similar trick that you're doing right now, just with inline policies except managed policies? Since there is no limit on the number of inline policies. Maybe change the logical ID of the inline policy based on the properties of the Table that can cause replacement? Would that work?

**Comment by jogold** [comment]
> Maybe change the logical ID of the inline policy based on the properties of the Table that can cause replacement? Would that work?

This should be possible but would restrict users from using tokens in those properties?

**Comment by skinny85** [comment]
Yeah, you're right, it wouldn't be as fool-proof as the current trick... I guess counting to 10 is the way to go here!

**Comment by jogold** [comment]
> Probably including the value of `timeout` in the UID is the way to go here.

Actually we cannot do this. For users already using a timeout if we include it in the uid it will impact the logical ID of the Lambda backing the custom resource. This means that the service token of the custom resource must be updated and this is not allowed in CF.

No problem for counting to 10 because we can maintain the existing uid (and thus logical ids) up to 10.

**Comment by reinismu** [comment]
Seems like I'm stuck at the old version for my project :/

**Comment by irobertson** [comment]
Just ran into this today; leaving comment to avoid issue auto-closing.

**Comment by oleg-slapdash** [comment]
Try setting `@aws-cdk/aws-iam:minimizePolicies` feature flag https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-iam.PolicyDocumentProps.html#minimize

**Comment by Tengda** [comment]
I have more than 10 tables with db replication and got the same error...

**Comment by FelixRelli** [comment]
Just ran into increased 20 limit. @aws-cdk/aws-iam:minimizePolicies does not help.

**Comment by tobiasviehweger** [comment]
Running into this as well... having to split my persistence into different stacks because of such an arbitrary low limit is extremely unfortunate.

**Comment by rix0rrr** [comment]
This issue was for the existing `Table` construct, which used custom resources to implement table replication. We no longer recommend the use of the `Table` construct.

Instead, the `TableV2` construct has been released in 2.95.1 (#27023) which maps to the `AWS::DynamoDB::GlobalTable` resource, has better support for replication and does not suffer from the issue described here. 

---

Be aware that there are additional deployment steps involved in a migration from `Table` to `TableV2`. You need to do a RETAIN deployment, a delete deployment, then change the code to use `TableV2` and then use `cdk import`. A link to a full guide will be posted once it is available. 

Here are some other resources to get you started (using `CfnGlobalTable` instead of `TableV2`) if you want to get going on the migration:

* [Migrating from Table to CfnGlobalTable](https://github.com/aws-samples/amazon-dynamodb-table-to-global-table-cdk), by Samaneh Utter, an AWS Senior Solutions Architect
* [AWS CDK and Amazon DynamoDB Global Tables](https://dev.to/aws-builders/aws-cdk-and-amazon-dynamodb-global-tables-3p0b), by Wojciech Matuszewski from Stedi
* [Migrating from Table to CfnGlobalTable](https://medium.com/@sharon.ad94/effortlessly-migrate-dynamodb-table-to-global-table-with-cloudformation-or-cdk-no-downtime-no-a1f62093d420), by Sharon Diskin from CloudBit
